File: typing_monkeys/individual.py
import random


class Individual:

    ALPHABET = 'abcdefghijklmnopqrstuvxywz '

    def __init__(self, genes):
        self.genes = genes
        self.fitness = None

    def calculate_fitness(self, target):
        score = 0
        for char in range(len(target)):
            if target[char] == self.genes[char]:
                score += 1
        self.fitness = score/len(target)

    def crossover(self, other):
        child_genes = []
        pivot = random.randrange(len(self.genes))
        return Individual(self.genes[:pivot] + other.genes[pivot:])

    def mutate(self, mutation_rate):
        new_genes = list(self.genes)
        for i in range(len(self.genes)):
            if random.random() < mutation_rate:
                new_genes[i] = random.choice(list(Individual.ALPHABET))
        self.genes = ''.join(new_genes)

File: typing_monkeys/ruby/individual.rb
class Individual
  attr_accessor :genes
  attr_accessor :fitness

  def initialize(genes, alphabet)
    @genes = genes
    @alphabet = alphabet
  end

  def calculate_fitness(target)
    score = 0
    @genes.each_with_index do |gene, index|
      score = score + 1 if target[index]==gene
    end
    @fitness = score/Float(@genes.size)
  end

  def crossover(other)
    pivot = (rand * (@genes.size-2)).floor
    Individual.new(@genes[0..pivot] + other.genes[(pivot+1)..(@genes.size-1)], @alphabet)
  end

  def mutate(ratio)
    (0..@genes.size-1).each do |i|
      @genes[i] = @alphabet[(rand * @alphabet.size).floor] if rand < ratio
    end
  end
end
